### PR TITLE
Add MPI cluster environment

### LINKED ISSUE #14014: ClusterEnvironment for Azure/MPI
## 🚀 Feature

Provide a ClusterEnvironment, either specific to Azure or a general one for MPI launchers.

### Motivation

Users have asked about this, indirectly, through issues like #13907 and #13639 where they struggled to get multi-node working on Azure/MPI clusters. If Lightning could auto-detect these clusters like we do for others, it would make it easier for these users to get Lightning to run there. There is also some PyTorch Lightning documentation on [Azure's side here](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-train-distributed-gpu#pytorch-lightning) that could look much cleaner if we provided this out of the box.

### Pitch

It would roughly look like this (sketch).

For a general MPI configuration:

```py
import os
from pytorch_lightning.plugins.environments import ClusterEnvironment


class OpenMPIClusterEnvironment(ClusterEnvironment):
    @property
    def creates_processes_externally(self) -> bool:
        """Return True if the cluster is managed (you don't launch processes yourself)"""
        return True

    def world_size(self) -> int:
        return int(os.environ["OMPI_COMM_WORLD_SIZE"])

    def set_world_size(self, size: int) -> None:
        pass

    def global_rank(self) -> int:
        return int(os.environ["OMPI_COMM_WORLD_RANK"])

    def set_global_rank(self, rank: int) -> None:
        pass

    def local_rank(self) -> int:
        return int(os.environ["OMPI_COMM_WORLD_LOCAL_RANK"])

    def node_rank(self) -> int:
        # this may not exist, defaulting to 0
        return int(os.environ.get("OMPI_COMM_WORLD_NODE_RANK", 0))

    def main_address(self) -> str:
        return os.environ["MASTER_ADDRESS"]

    def main_port(self) -> int:
        return int(os.environ["MASTER_PORT"])

    @staticmethod
    def detect() -> bool:
        return "OMPI_COMM_WORLD_SIZE" in os.environ


trainer = Trainer(plugins=[OpenMPIClusterEnvironment()])
```

With Azure, we would have to incorporate specific environment variables like 

```
AZ_BATCH_MASTER_NODE
AZ_BATCHAI_MPI_MASTER_NODE
```

when determining the main server address.

### Alternatives

Let users implement these environments from scratch.
We will provide an example/template in the docs that users can copy paste. The docs would have a specific section about Azure.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @borda @awaelchli @ananthsub

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds an environment for MPI.

Fixes #14014

TODO:
- [x] Tests
- [x] Proper detection
- [x] Integration into Lightning connector
- [x] Validation on real cluster with MPI

Inspired by [this utility function](https://github.com/coreyjadams/CosmicTagger/blob/b36c0f89d57769437ad5d8bd42af1c5b7ad3be40/src/utils/core/mpi_utils.py#L5-L89).

Co-author: Corey Adams @coreyjadams

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Comment by coreyjadams** [comment]
Some comments: 
- this works out of the box using our pytorch + MPI on ALCF supercomputers, at least for pytorch DDP. So, it is tested on our system.
- Some warnings emitted by pytorch get really verbose as the scale goes up.  All but one can be globally disabled in python in user code, though it'd be nice if there was a way to apply `@rank_zero_only` to warnings.
- The only warning that isn't suppressible in user would be squashed by PR #16611 .

Thanks for doing this!

**Comment by awaelchli** [comment]
@coreyjadams Awesome, thanks for confirming this!

> Some warnings emitted by pytorch get really verbose as the scale goes up. 

Do you specifically mean pytorch, or messages from lightning? If you could provide me an example message that would be great. We generally log our warnings and messages in Lightning only on rank 0.

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Comment by MaximilienLC** [comment]
Hi! Thanks for the great feature. Any tips on how to use this for a multi-node MPI cluster?
